docs/05-rfcs/2023-01-11-sdk-spec.md

```diff
@@ -0,0 +1,1070 @@
+jsii poses restrictions on language features that cannot be idiomatically represented in target languages, and encourages good practices for object-oriented design.
+Features that are specific to Wing (such as inflight functions) may not be supported in other jsii languages.
+
+## Concepts used throughout the SDK
+
+### Serializable
+
+> Reqtag: `sdk:serializable`
+<span id="sdk:serializable"/>
+
+A `Serializable` is any immutable value that can be serialized and sent over the wire.
+This includes all primitive types, as well as any immutable collection types of `Serializable`s (such as `List`s and `Map`s).
+
+This RFC uses the term `Serializable` as a placeholder type in several parts of the API, but note that it may be replaced in the future with more specific types such as `Json` or `Blob` (see [this issue](https://github.com/winglang/wing/issues/185)).
+
+### Paginated APIs
+
+> Reqtag: `sdk:pagination`
+<span id="sdk:pagination"/>
+
+Some APIs return a list of results that may be too large to fit in memory, or too large to fetch from the cloud all at once.
+In these cases, APIs can return an `Iterator` object.
+An `Iterator` is an object with a `next()` method and a `has_next()` method, which return the next page of results and whether there are more results to fetch, respectively.
+The `Iterator` object also implements the [async iterator protocol in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols), so it can be used within `for...of` loops in TypeScript or in Wing.
+
+### Event handlers
+
+Some resources can automatically emit events as a result of an action or a change to their state.
+These events are "fire-and-forget" notifications which may not always be delivered in order, and which don't expect a response to be sent back by the listener(s).
+
+Resources that emit events in a "fire-and-forget" fashion should make their events listenable through a method named `on_<event>`.
+
+> Reqtag: `sdk:on-event-name`
+<span id="sdk:on-event-name"/>
+
+For example, a `cloud.Bucket` can emit events whenever an object is uploaded, and these events can be listened to by calling `on_upload` on the bucket:
+
+```ts
+// wing
+let bucket = new cloud.Bucket();
+bucket.on_upload(inflight (file: str) => {
+  log.info("File uploaded: " + file);
+});
+```
+
+The argument to an `on_<event>` method is called an event handler.
+An event handler is any resource that implements the `EventHandler<T>` interface, which contains a single method named `handle` that accepts an event of type `T`:
+
+```ts
+// wing
+
+interface EventHandler<T> {
+  inflight handle(event: T);
+}
+```
+
+<!--
+We could also decide the should be named `handle_event` or `invoke` or `run` or `call` or something else.
```

Comment:
Ok - updated the description to avoid this general `EventHandler` interface and allude to some kind of structural typing. I didn't add too much detail since I felt it's more of a compiler/language detail